crates/anemo-tower/src/rate_limit.rs

```diff
@@ -0,0 +1,195 @@
+        DefaultClock,
+        NoOpMiddleware<<DefaultClock as Clock>::Instant>,
+    >,
+>;
+
+/// [`Layer`] for adding a per-peer rate limit to inbound requests.
+///
+/// See the [module docs](super::rate_limit) for more details.
+#[derive(Clone, Debug)]
+pub struct RateLimitLayer {
+    pub limiter: SharedRateLimiter,
+}
+
+impl RateLimitLayer {
+    /// Creates a new [`RateLimitLayer`].
+    pub fn new(quota: governor::Quota) -> Self {
+        RateLimitLayer {
+            limiter: Arc::new(RateLimiter::keyed(quota)),
+        }
+    }
+}
+
+impl<S> Layer<S> for RateLimitLayer {
+    type Service = RateLimit<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        RateLimit {
+            inner,
+            limiter: self.limiter.clone(),
+        }
+    }
+}
+
+/// Middleware for adding a per-peer rate limit to inbound requests.
+///
+/// See the [module docs](super::rate_limit) for more details.
+#[derive(Clone, Debug)]
+pub struct RateLimit<S> {
+    pub inner: S,
+    pub limiter: SharedRateLimiter,
+}
+
+impl<S> RateLimit<S> {
+    /// Creates a new [`RateLimit`].
+    pub fn new(inner: S, quota: governor::Quota) -> Self {
+        Self {
+            limiter: Arc::new(RateLimiter::keyed(quota)),
+            inner,
+        }
+    }
+
+    /// Gets a reference to the underlying service.
+    pub fn inner_ref(&self) -> &S {
+        &self.inner
+    }
+
+    /// Gets a mutable reference to the underlying service.
+    pub fn inner_mut(&mut self) -> &mut S {
+        &mut self.inner
+    }
+
+    /// Consumes `self`, returning the underlying service.
+    pub fn into_inner(self) -> S {
+        self.inner
+    }
+
+    /// Returns a new [`Layer`] that wraps services with a `RateLimit` middleware.
+    ///
+    /// [`Layer`]: tower::layer::Layer
+    pub fn layer(quota: governor::Quota) -> RateLimitLayer {
+        RateLimitLayer {
+            limiter: Arc::new(RateLimiter::keyed(quota)),
+        }
+    }
+}
+
+impl<ResBody, ReqBody, S> Service<Request<ReqBody>> for RateLimit<S>
+where
+    S: Service<Request<ReqBody>, Response = Response<ResBody>, Error = anemo::rpc::Status>
+        + 'static
+        + Clone
+        + Send,
+    <S as Service<Request<ReqBody>>>::Future: Send,
+    ReqBody: 'static + Send + Sync,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;
+
+    #[inline]
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Request<ReqBody>) -> Self::Future {
+        let limiter = self.limiter.clone();
+        let mut inner = self.inner.clone();
+
+        let fut = async move {
+            let peer_id = req.peer_id().ok_or_else(|| {
+                anemo::rpc::Status::internal("rate limiter missing request PeerId")
+            })?;
+            limiter.until_key_ready(peer_id).await;
```

Comment:
So i suppose there are two ways to do this, we can either wait till we have quota or we can return an error early. I don't have a strong opinion on which we do, but it may be worth adding some documentation to the module indicating the behavior.

Reply:
Updated this to make it an option